content/blog/2021-12-07-ml-experiment-versioning.md

```diff
@@ -0,0 +1,240 @@
+- **Distributed experiments**: Organize locally and choose what to share,
+  reusing your existing repo setup.
+
+```
++---------------------+    +------------------+
+|                     |    |                  |
+| Experiment Tracking |    |  Code Versioning |
+|                     |    |                  |
++---------------------+    +---------/--------+
+            \                       /
+             \                     /
+              \-                 /-
+                \               /
+                 \             /
+            +-----\------------------+
+            |                        |
+            | Experiment Versioning  |
+            |                        |
+            +------------------------+
+```
+
+# ML Experiments as Code
+
+Experiment versioning treats experiments as code. Experiment tracking tools log
+parameters, metrics, and artifacts to a database through API calls. In
+experiment versioning, they are saved in files.
+
+You can choose your own file formats and paths, which you can configure in DVC:
+
+```dvc
+$ dvc exp init -i
+This command will guide you to set up a default stage in dvc.yaml.
+See https://dvc.org/doc/user-guide/project-structure/pipelines-files.
+
+DVC assumes the following workspace structure:
+├── data
+├── metrics.json
+├── models
+├── params.yaml
+├── plots
+└── src
+
+Command to execute: python src/train.py
+Path to a code file/directory [src, n to omit]: src/train.py
+Path to a data file/directory [data, n to omit]: data/images/
+Path to a model file/directory [models, n to omit]:
+Path to a parameters file [params.yaml, n to omit]:
+Path to a metrics file [metrics.json, n to omit]:
+Path to a plots file/directory [plots, n to omit]: logs.csv
+```
+
+This might mean a little more work to set up your code to read parameters from a
+file or write metrics to another file. To track artifacts like data and models
+in your repo, DVC adds
+[data versioning](https://dvc.org/doc/start/data-and-model-versioning) support.
+
+Once you set up your repo in this structure, you start to see the benefits of
+this approach. Experiment meta-information lives in readable files that are
+always available, and your code can stay clean. You can read, save, and version
+your meta-information:
+
+```dvc
+$ cat params.yaml
+train:
+  epochs: 10
+model:
+  conv_units: 128
+```
+
+```dvc
+$ cat metrics.json
+{"loss": 0.24310708045959473, "acc": 0.9182999730110168}
+```
+
+You can see what changed in parameters, code, or anything else:
+
+```dvc
+$ git diff HEAD~1 -- params.yaml
+diff --git a/params.yaml b/params.yaml
+index baad571a2..57d098495 100644
+--- a/params.yaml
++++ b/params.yaml
+@@ -1,5 +1,5 @@
+ train:
+   epochs: 10
+-model:
+-  conv_units: 16
++model:
++  conv_units: 128
+```
+
+With DVC, you can compare lots of experiments at a glance:
+
+```dvc
+$ dvc exp show
+┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
```

Comment:
For me the table is broken a bit
<img width="1115" alt="Screen Shot 2021-11-30 at 8 16 52 PM" src="https://user-images.githubusercontent.com/10081867/144170913-57a760bd-36f6-464e-907e-d47d7640dda0.png">

Reply:
Do you mean the "Experiment row" looking not quite aligned? Here's how it looks for me:

<img width="726" alt="Screen Shot 2021-12-01 at 8 49 25 PM" src="https://user-images.githubusercontent.com/2308172/144342821-efa03a42-f2f1-4fd4-83a6-1620fa853231.png">

Reply:
I see extra spaces in rows. All the experiment IDs should be one right after another. But I see a space line.